lib/krikri/enricher.rb

```diff
@@ -0,0 +1,132 @@
+  #       input_fields: [{sourceResource: :title}]
+  #     },
+  #     'Krikri::Enrichments::StripWhitespace' => {
+  #       input_fields: [{sourceResource: :title}]
+  #     }
+  #   }
+  #   Krikri::Enricher.enqueue({
+  #     generator_uri: 'http://ldp.local.dp.la/ldp/activity/3',
+  #     chain: chain
+  #   })
+  #
+  # @see Krikri::SoftwareAgent#enqueue
+  #
+  class Enricher
+    include SoftwareAgent
+
+    attr_reader :chain, :generator_uri
+
+    def self.queue_name
+      :enrichment
+    end
+
+    def initialize(opts = {})
+      @generator_uri = RDF::URI(opts.fetch(:generator_uri))
+      @chain = deep_sym(opts.fetch(:chain) { {} })
+    end
+
+    def run(activity_uri = nil)
+      log :info, 'enricher is running'
+      # see TODO below
+      target_aggregations.each do |agg|
+        chain_enrichments!(agg)
+        agg << RDF::Statement(agg, RDF::PROV.wasGeneratedBy, activity_uri) if
+          activity_uri
+        agg.save
+      end
+    rescue => e
+      Rails.logger.error("Enrichment error: #{e.message}\n#{e.backtrace}")
+    ensure
+      log :info, 'enricher is done'
+    end
+
+    # TODO:  remove this when the current topic branch that introduces the
+    # EntityConsumer mixin has been merged.
+    def target_aggregations
+      query = Krikri::ProvenanceQueryClient.find_by_activity(generator_uri)
+      query.execute.lazy.flat_map do |solution|
+        agg = DPLA::MAP::Aggregation.new(solution.record.to_s)
+        agg.get
+        agg
+      end
+    end
+
+    def chain_enrichments!(agg)
+      chain.keys.each do |e|
+        enrichment = e.to_s.constantize.new
+        if enrichment.is_a? Krikri::FieldEnrichment
+          agg = do_field_enrichment(agg, enrichment, chain[e])
+        else
+          agg = do_basic_enrichment(agg, enrichment, chain[e])
+        end
+      end
+    end
+
+    private
+
+    ##
+    # Perform a default enrichment, using Enrichment#enrichment or a derived
+    # class that expects the same arguments.
+    #
+    # @param agg [DPLA::MAP::Aggregation]
+    # @param enrichment [Krikri::Enrichment]
+    # @param options [Hash]
+    #
+    def do_basic_enrichment(agg, enrichment, options)
+      enrichment.enrich(
+        agg, options[:input_fields], options[:output_fields]
+      )
+    end
+
+    ##
+    # Perform a FieldEnrichment enrichment on the given aggregation.
+    #
+    # With FieldEnrichment#enrich, the input_fields option parameter is passed
+    # as a variable arguments list
+    #
+    # @param agg [DPLA::MAP::Aggregation]
+    # @param enrichment [Krikri::FieldEnrichment]
+    # @param options [Hash]  Hash with :input_fields containing variable
+    #                        arguments list
+    #
+    def do_field_enrichment(agg, enrichment, options)
+      enrichment.enrich(agg, *options[:input_fields])
+    end
+
+    ##
+    # Transform the given hash recursively by turning all of its string keys
+    # and values into symbols.
+    #
```

Comment:
Would [`Hash#deep_symbolize_keys`](http://api.rubyonrails.org/classes/Hash.html#method-i-deep_symbolize_keys) from `ActiveSupport`'s core extensions for `Hash` work?